README.md
# SRMCategory

[![CI Status](http://img.shields.io/travis/SongRanMark/SRMCategory.svg?style=flat)](https://travis-ci.org/SongRanMark/SRMCategory)
[![Version](https://img.shields.io/cocoapods/v/SRMCategory.svg?style=flat)](http://cocoapods.org/pods/SRMCategory)
[![License](https://img.shields.io/cocoapods/l/SRMCategory.svg?style=flat)](http://cocoapods.org/pods/SRMCategory)
[![Platform](https://img.shields.io/cocoapods/p/SRMCategory.svg?style=flat)](http://cocoapods.org/pods/SRMCategory)

SRMCategory 是一个为 iOS 开发提供通用 Category 集合的类库。对 Foundation 和 UIKit 原生库中的类提供了一些新的方法，以提高开发的效率和代码的简洁度。

## 安装

```ruby
pod "SRMCategory"
```

## 使用

### Category 列表

在相应文件中导入需要的 Category 即可使用其中的方法，每个 Category 中所包含的方法，请查阅其 .h 文件。

- Foundation

	- NSBundle(SRMExtension)
	- NSString(SRMExtension)
	- NSURL(SRMExtension)
	- NSUserDefaults(SRMExtension)

- UIKit

	- UIColor(SRMExtension)
	- UIScreen(SRMExtension)
	- UIView(SRMExtension)
	- UIView(SRMConstraint)
	- UIView(SRMFrame)
	- UIViewController(SRMExtension)

### 分类规则

初始每一个类创建的 Category 都为 SRMExtension，之后若某一主题的相关方法有很多，则会单独新建一个对应主题的分类，如 UIView(SRMConstraint)，UIView(SRMFrame)。

### 测试

所有的分类方法都通过 Example 目录下的 SRMCategory 工程进行单元测试，同时也可以通过 test case 查看分类方法如何使用。新添加的分类方法也必须配套添加对应的测试方法，以保证这些分类方法的稳定性。

## 联系我

S.R, firecrackerinlane2@gmail.com

## 授权许可

**MIT**


CHANGELOG.md
+ Version 0.1.0: 初始化 Category 集的第一个版本.


Example/Tests/Tests.m
//
//  SRMCategoryTests.m
//  SRMCategoryTests
//
//  Created by githubsr on 06/26/2017.
//  Copyright (c) 2017 githubsr. All rights reserved.
//

@import XCTest;

@interface Tests : XCTestCase

@end

@implementation Tests

- (void)setUp
{
    [super setUp];
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown
{
    // Put teardown code here. This method is called after the invocation of each test method in the class.
    [super tearDown];
}

- (void)testExample
{
    XCTFail(@"No implementation for \"%s\"", __PRETTY_FUNCTION__);
}

@end



Example/Tests/UIView_Frame_Tests.m
//
//  UIColor_Extension_Tests.m
//  Category Demo
//
//  Created by marksong on 07/06/2017.
//  Copyright © 2017 SR. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <SRMCategory/SRMCategory-umbrella.h>

@interface UIView_Frame_Tests : XCTestCase

@end

@implementation UIView_Frame_Tests

- (void)test {
    CGFloat originX = 5, originY = 10, width = 50, height = 100;
    UIView *view = [[UIView alloc] initWithFrame:CGRectMake(originX, originY, width, height)];
    XCTAssert(view.originX == originX);
    XCTAssert(view.originY == originY);
    XCTAssert(CGPointEqualToPoint(view.origin, CGPointMake(originX, originY)));
    XCTAssert(view.width == width);
    XCTAssert(view.height == height);
    XCTAssert(CGSizeEqualToSize(view.size, CGSizeMake(width, height)));
    CGFloat newOriginX = 10, newOriginY = 5, newWidth = 100, newHeight = 50;
    view.originX = newOriginX;
    view.originY = newOriginY;
    view.width = newWidth;
    view.height = newHeight;
    XCTAssert(view.originX == newOriginX);
    XCTAssert(view.originY == newOriginY);
    XCTAssert(CGPointEqualToPoint(view.origin, CGPointMake(newOriginX, newOriginY)));
    XCTAssert(view.width == newWidth);
    XCTAssert(view.height == newHeight);
    XCTAssert(CGSizeEqualToSize(view.size, CGSizeMake(newWidth, newHeight)));
}

@end


Example/Tests/NSURL_Extension_Tests.m
//
//  NSURL_Extension_Tests.m
//  Category Demo
//
//  Created by marksong on 08/06/2017.
//  Copyright © 2017 SR. All rights reserved.
//

#import <XCTest/XCTest.h>
#import <SRMCategory/SRMCategory-umbrella.h>

static NSString *const kDomain = @"http://example.com";

@interface NSURL_Extension_Tests : XCTestCase

@end

@implementation NSURL_Extension_Tests

#pragma mark - URLByAppendingQueryItemWithName_value

- (void)test_URLByAppendingQueryItemWithName_value_noQueryItem {
    NSURL *defaultURL = [NSURL URLWithString:kDomain];
    NSURL *resultURL = [defaultURL srm_URLByAppendingQueryItemWithName:@"key" value:@"value"];
    NSString *resultURLString = [kDomain stringByAppendingString:@"?key=value"];
    XCTAssert([resultURL.absoluteString isEqualToString:resultURLString]);
}

- (void)test_URLByAppendingQueryItemWithName_value_haveQueryItem {
    NSString *defaultURLString = [kDomain stringByAppendingString:@"?key1=value1"];
    NSURL *defaultURL = [NSURL URLWithString:defaultURLString];
    NSURL *resultURL = [defaultURL srm_URLByAppendingQueryItemWithName:@"key2" value:@"value2"];
    NSString *resultURLString = [defaultURLString stringByAppendingString:@"&key2=value2"];
    XCTAssert([resultURL.absoluteString isEqualToString:resultURLString]);
}

- (void)test_URLByAppendingQueryItemWithName_value_nilQueryItem {
    NSURL *defaultURL = [NSURL URLWithString:kDomain];
    NSURL *resultURL = [defaultURL srm_URLByAppendingQueryItemWithName:nil value:nil];
    XCTAssert([resultURL.absoluteString isEqualToString:kDomain]);
}

- (void)test_URLByAppendingQueryItemWithName_value_emptyQueryItem {
    NSURL *defaultURL = [NSURL URLWithString:kDomain];
    NSURL *resultURL = [defaultURL srm_URLByAppendingQueryItemWithName:@"" value:@""];
    XCTAssert([resultURL.absoluteString isEqualToString:kDomain]);
}

- (void)test_URLByAppendingQueryItemWithName_value_invalidQueryItem {
    NSURL *defaultURL = [NSURL URLWithString:kDomain];
    NSURL *resultURL = [defaultURL srm_URLByAppendingQueryItemWithName:@"key" value:@""];
    XCTAssert([resultURL.absoluteString isEqualToString:kDomain]);
}

- (void)test_URLByAppendingQueryItemWithName_value_chineseQueryItem {
    NSURL *defaultURL = [NSURL URLWithString:kDomain];
    NSURL *resultURL = [defaultURL srm_URLByAppendingQueryItemWithName:@"键" value:@"值"];
    NSString *unencodeURLString = [kDomain stringByAppendingString:@"?键=值"];
    NSCharacterSet *characterSet = [NSCharacterSet URLQueryAllowedCharacterSet];
    NSString *resultURLString = [unencodeURLString stringByAddingPercentEncodingWithAllowedCharacters:characterSet];
    XCTAssert([resultURL.absoluteString isEqualToString:resultURLString]);
}

#pragma mark - URLByAppendingQueryItems

- (void)test_URLByAppendingQueryItems_normal {
    NSURL *defaultURL = [NSURL URLWithString:kDomain];
    NSDictionary *queryItems = @{
                                 @"key" : @"value",
                                 @"键" : @"值"
                                 };
    NSURL *resultURL = [defaultURL srm_URLByAppendingQueryItems:queryItems];
    NSString *unencodeURLString = [kDomain stringByAppendingString:@"?key=value&键=值"];
    NSCharacterSet *characterSet = [NSCharacterSet URLQueryAllowedCharacterSet];
    NSString *resultURLString = [unencodeURLString stringByAddingPercentEncodingWithAllowedCharacters:characterSet];
    XCTAssert([resultURL.absoluteString isEqualToString:resultURLString]);
}

- (void)test_URLByAppendingQueryItems_nil {
    NSURL *defaultURL = [NSURL URLWithString:kDomain];
    NSURL *resultURL = [defaultURL srm_URLByAppendingQueryItems:nil];
    XCTAssert([resultURL.absoluteString isEqualToString:kDomain]);
}

- (void)test_URLByAppendingQueryItems_empty {
    NSURL *defaultURL = [NSURL URLWithString:kDomain];
    NSDictionary *queryItems = @{
                                 @"" : @"",
                                 @"key" : @""
                                 };
    NSURL *resultURL = [defaultURL srm_URLByAppendingQueryItems:queryItems];
    XCTAssert([resultURL.absoluteString isEqualToString:kDomain]);
}

@end


Example/SRMCategory/SRMAppDelegate.m
//
//  SRMAppDelegate.m
//  SRMCategory
//
//  Created by githubsr on 06/26/2017.
//  Copyright (c) 2017 githubsr. All rights reserved.
//

#import "SRMAppDelegate.h"

@implementation SRMAppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
    // Override point for customization after application launch.
    return YES;
}

- (void)applicationWillResignActive:(UIApplication *)application
{
    // Sent when the application is about to move from active to inactive state. This can occur for certain types of temporary interruptions (such as an incoming phone call or SMS message) or when the user quits the application and it begins the transition to the background state.
    // Use this method to pause ongoing tasks, disable timers, and throttle down OpenGL ES frame rates. Games should use this method to pause the game.
}

- (void)applicationDidEnterBackground:(UIApplication *)application
{
    // Use this method to release shared resources, save user data, invalidate timers, and store enough application state information to restore your application to its current state in case it is terminated later.
    // If your application supports background execution, this method is called instead of applicationWillTerminate: when the user quits.
}

- (void)applicationWillEnterForeground:(UIApplication *)application
{
    // Called as part of the transition from the background to the inactive state; here you can undo many of the changes made on entering the background.
}

- (void)applicationDidBecomeActive:(UIApplication *)application
{
    // Restart any tasks that were paused (or not yet started) while the application was inactive. If the application was previously in the background, optionally refresh the user interface.
}

- (void)applicationWillTerminate:(UIApplication *)application
{
    // Called when the application is about to terminate. Save data if appropriate. See also applicationDidEnterBackground:.
}

@end


Example/SRMCategory/SRMViewController.h
//
//  SRMViewController.h
//  SRMCategory
//
//  Created by githubsr on 06/26/2017.
//  Copyright (c) 2017 githubsr. All rights reserved.
//

@import UIKit;

extern NSString *const kSuperViewIdentifier;
extern NSString *const kCenterViewIdentifier;
extern NSString *const kCenterConstantViewIdentifier;
extern NSString *const kSizeViewIdentifier;
extern NSString *const kEqualSizeViewIdentifier;
extern NSString *const kEqualEdgeViewIdentifier;
extern NSString *const kEdgeConstantViewIdentifier;
extern NSString *const kSubviewIdentifier;
extern NSString *const kNextToXViewIdentifier;
extern NSString *const kNextToYViewIdentifier;
extern const CGFloat kCenterViewConstantX;
extern const CGFloat kCenterViewConstantY;
extern const CGFloat kViewWidth;
extern const CGFloat kViewHeight;
extern const CGFloat kEdgeInsetTop;
extern const CGFloat kEdgeInsetBottom;
extern const CGFloat kEdgeInsetLeading;
extern const CGFloat kEdgeInsetTrailing;

@interface SRMViewController : UIViewController

@end


Example/SRMCategory/SRMViewController.m
//
//  SRMViewController.m
//  SRMCategory
//
//  Created by githubsr on 06/26/2017.
//  Copyright (c) 2017 githubsr. All rights reserved.
//

#import "SRMViewController.h"
#import <SRMCategory/SRMCategory-umbrella.h>

NSString *const kSuperViewIdentifier = @"kSuperViewIdentifier";
NSString *const kCenterViewIdentifier = @"kCenterViewIdentifier";
NSString *const kCenterConstantViewIdentifier = @"kCenterConstantViewIdentifier";
NSString *const kSizeViewIdentifier = @"kSizeViewIdentifier";
NSString *const kEqualSizeViewIdentifier = @"kEqualSizeViewIdentifier";
NSString *const kEqualEdgeViewIdentifier = @"kEqualEdgeViewIdentifier";
NSString *const kEdgeConstantViewIdentifier = @"kEdgeConstantViewIdentifier";
NSString *const kSubviewIdentifier = @"kSubviewIdentifier";
NSString *const kNextToXViewIdentifier = @"kNextToXViewIdentifier";
NSString *const kNextToYViewIdentifier = @"kNextToYViewIdentifier";
const CGFloat kCenterViewConstantX = 10;
const CGFloat kCenterViewConstantY = 10;
const CGFloat kViewWidth = 50;
const CGFloat kViewHeight = 50;
const CGFloat kEdgeInsetTop = 5;
const CGFloat kEdgeInsetBottom = 5;
const CGFloat kEdgeInsetLeading = 5;
const CGFloat kEdgeInsetTrailing = 5;

@interface SRMViewController ()

@end

@implementation SRMViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.accessibilityIdentifier = kSuperViewIdentifier;
    [self addCenterView];
    [self addCenterConstantView];
    [self addSizeView];
    [self addEqualSizeView];
    [self addEqualEdgeView];
    [self addEdgeConstantView];
    [self addNextToView];
}

- (void)addCenterView {
    UIView *view = [[UIView alloc] init];
    [self.view addSubview:view];
    view.accessibilityIdentifier = kCenterViewIdentifier;
    [view srm_addSizeConstraints:CGSizeMake(kViewWidth, kViewHeight)];
    [view srm_addCenterConstraintsEqualToView:self.view];
}

- (void)addCenterConstantView {
    UIView *view = [[UIView alloc] init];
    [self.view addSubview:view];
    view.accessibilityIdentifier = kCenterConstantViewIdentifier;
    [view srm_addSizeConstraints:CGSizeMake(kViewWidth, kViewHeight)];
    [view srm_addCenterConstraintsToView:self.view offset:UIOffsetMake(kCenterViewConstantX, kCenterViewConstantY)];
}

- (void)addSizeView {
    UIView *view = [[UIView alloc] init];
    [self.view addSubview:view];
    view.accessibilityIdentifier = kSizeViewIdentifier;
    [view srm_addSizeConstraints:CGSizeMake(kViewWidth, kViewHeight)];
    [view srm_addCenterConstraintsEqualToView:self.view];
}

- (void)addEqualSizeView {
    UIView *view = [[UIView alloc] init];
    [self.view addSubview:view];
    view.accessibilityIdentifier = kEqualSizeViewIdentifier;
    [view srm_addSizeConstraintsEqualToView:self.view];
    [view srm_addCenterConstraintsEqualToView:self.view];
}

- (void)addEqualEdgeView {
    UIView *view = [[UIView alloc] init];
    [self.view addSubview:view];
    view.accessibilityIdentifier = kEqualEdgeViewIdentifier;
    [view srm_addEdgesConstraintsEqualToView:self.view];
}

- (void)addEdgeConstantView {
    UIView *view = [[UIView alloc] init];
    [self.view addSubview:view];
    view.accessibilityIdentifier = kEdgeConstantViewIdentifier;
    UIEdgeInsets edgeInsets = UIEdgeInsetsMake(kEdgeInsetTop, kEdgeInsetLeading, kEdgeInsetBottom, kEdgeInsetTrailing);
    [view srm_addEdgesConstraintsToView:self.view insets:edgeInsets];
}

- (void)addNextToView {
    UIView *subview = [[UIView alloc] init];
    [self.view addSubview:subview];
    subview.accessibilityIdentifier = kSubviewIdentifier;
    [subview srm_addSizeConstraints:CGSizeMake(kViewWidth, kViewHeight)];
    [subview srm_addCenterConstraintsEqualToView:self.view];
    UIView *nextToXView = [[UIView alloc] init];
    [self.view addSubview:nextToXView];
    nextToXView.accessibilityIdentifier = kNextToXViewIdentifier;
    [nextToXView srm_addSizeConstraints:CGSizeMake(kViewWidth, kViewHeight)];
    [nextToXView srm_addCenterConstraintEqualToViewVertically:subview];
    [nextToXView srm_addConstraintNextToViewHorizontally:subview];
    UIView *nextToYView = [[UIView alloc] init];
    [self.view addSubview:nextToYView];
    nextToYView.accessibilityIdentifier = kNextToYViewIdentifier;
    [nextToYView srm_addSizeConstraints:CGSizeMake(kViewWidth, kViewHeight)];
    [nextToYView srm_addCenterConstraintEqualToViewHorizontally:subview];
    [nextToYView srm_addConstraintNextToViewVertically:subview];
}

@end
